File: android/engine/src/main/kotlin/com/knitstudio/engine/ProjectPlan.kt
package com.knitstudio.engine

/** A single headline number shown at the top of a plan. */
data class PlanFact(
    val label: String,
    val value: String,
    val detail: String? = null,
)

/** One instruction the knitter follows. */
data class PlanStep(
    val text: String,
    /** Stitch count in play once this step is finished. */
    val stitchCount: Int? = null,
    /** Rows or rounds this step occupies. */
    val rows: Int? = null,
    /** Set when the step matters enough to call out (a heel turn, a split). */
    val isMilestone: Boolean = false,
)

/** A named run of steps — cuff, yoke, heel, crown. */
data class PlanSection(
    val name: String,
    val detail: String? = null,
    val steps: List<PlanStep>,
) {
    val totalRows: Int get() = steps.sumOf { it.rows ?: 0 }
}

/** Everything the calculator works out for one project. */
data class ProjectPlan(
    val title: String,
    val gauge: Gauge,
    val facts: List<PlanFact>,
    val sections: List<PlanSection>,
    /** Stitch blocks feeding the yarn estimate. */
    val blocks: List<StitchBlock>,
    val notes: List<String> = emptyList(),
    /** Things that are off: gauge implausible, measurements inconsistent. */
    val warnings: List<String> = emptyList(),
) {
    val totalStitches: Double get() = blocks.sumOf { it.stitches }
    val totalRows: Int get() = sections.sumOf { it.totalRows }

    fun metres(): Double = YarnEstimator.metres(blocks, gauge)

    fun fact(label: String): String? = facts.firstOrNull { it.label == label }?.value

    /** Flat plain-text rendering, used for sharing and export. */
    fun plainText(units: UnitSystem): String {
        val lines = mutableListOf(title, "=".repeat(title.length), "")
        lines += "Gauge: ${gauge.describe(units)}"
        lines += ""
        facts.forEach { fact ->
            lines += if (fact.detail != null) {
                "${fact.label}: ${fact.value}  (${fact.detail})"
            } else {
                "${fact.label}: ${fact.value}"
            }
        }
        sections.forEach { section ->
            lines += ""
            lines += section.name.uppercase()
            section.detail?.let { lines += it }
            section.steps.forEach { step ->
                var line = "  • ${step.text}"
                step.stitchCount?.let { line += " ($it sts)" }
                lines += line
            }
        }
        if (notes.isNotEmpty()) {
            lines += ""
            lines += "NOTES"
            notes.forEach { lines += "  • $it" }
        }
        if (warnings.isNotEmpty()) {
            lines += ""
            lines += "CHECK THESE"
            warnings.forEach { lines += "  ! $it" }
        }
        return lines.joinToString("\n")
    }
}

/** Common body measurements a project is calculated from. Stored in cm. */
data class BodyMeasurements(
    val chest: Double = 100.0,
    val headCircumference: Double = 56.0,
    val neckCircumference: Double = 38.0,
    val upperArm: Double = 34.0,
    val wrist: Double = 20.0,
    val bodyLength: Double = 62.0,
    val sleeveLength: Double = 45.0,
    val footCircumference: Double = 23.0,
    val footLength: Double = 24.0,
    val handCircumference: Double = 20.0,
    val handLength: Double = 19.0,
) {
    companion object {
        /** Standard adult sizes, so a knitter who has not measured can start. */
        val sizePresets: List<Pair<String, BodyMeasurements>> = listOf(
            "Child 6–8" to BodyMeasurements(
                chest = 66.0, headCircumference = 51.0, neckCircumference = 30.0, upperArm = 22.0,
                wrist = 15.0, bodyLength = 42.0, sleeveLength = 33.0,
                footCircumference = 17.0, footLength = 18.0,
                handCircumference = 15.0, handLength = 14.0,
            ),
            "Adult XS" to BodyMeasurements(
                chest = 82.0, headCircumference = 54.0, neckCircumference = 34.0, upperArm = 27.0,
                wrist = 16.0, bodyLength = 56.0, sleeveLength = 43.0,
                footCircumference = 20.0, footLength = 22.0,
                handCircumference = 17.0, handLength = 17.0,
            ),
            "Adult S" to BodyMeasurements(
                chest = 90.0, headCircumference = 55.0, neckCircumference = 36.0, upperArm = 30.0,
                wrist = 17.0, bodyLength = 59.0, sleeveLength = 44.0,
                footCircumference = 21.0, footLength = 23.0,
                handCircumference = 18.0, handLength = 18.0,
            ),
            "Adult M" to BodyMeasurements(),
            "Adult L" to BodyMeasurements(
                chest = 110.0, headCircumference = 57.0, neckCircumference = 40.0, upperArm = 37.0,
                wrist = 21.0, bodyLength = 65.0, sleeveLength = 46.0,
                footCircumference = 25.0, footLength = 26.0,
                handCircumference = 22.0, handLength = 20.0,
            ),
            "Adult XL" to BodyMeasurements(
                chest = 122.0, headCircumference = 58.0, neckCircumference = 42.0, upperArm = 41.0,
                wrist = 22.0, bodyLength = 68.0, sleeveLength = 47.0,
                footCircumference = 27.0, footLength = 27.0,
                handCircumference = 24.0, handLength = 21.0,
            ),
        )
    }
}

enum class Difficulty(val displayName: String, val order: Int) {
    BEGINNER("Beginner", 0),
    EASY("Easy", 1),
    INTERMEDIATE("Intermediate", 2),
    ADVANCED("Advanced", 3),
}

File: ios-knitstudio/KnitStudio/Engine/ProjectPlan.swift
import Foundation

/// A single headline number shown at the top of a plan.
struct PlanFact: Identifiable, Equatable {
    var id: UUID = UUID()
    var label: String
    var value: String
    var detail: String?

    init(_ label: String, _ value: String, detail: String? = nil) {
        self.label = label
        self.value = value
        self.detail = detail
    }
}

/// One instruction the knitter follows.
struct PlanStep: Identifiable, Equatable {
    var id: UUID = UUID()
    var text: String
    /// Stitch count in play once this step is finished.
    var stitchCount: Int?
    /// Rows or rounds this step occupies.
    var rows: Int?
    /// Set when the step matters enough to call out (a heel turn, a split).
    var isMilestone: Bool = false

    init(_ text: String, stitchCount: Int? = nil, rows: Int? = nil, isMilestone: Bool = false) {
        self.text = text
        self.stitchCount = stitchCount
        self.rows = rows
        self.isMilestone = isMilestone
    }
}

/// A named run of steps — cuff, yoke, heel, crown.
struct PlanSection: Identifiable, Equatable {
    var id: UUID = UUID()
    var name: String
    var detail: String?
    var steps: [PlanStep]

    init(_ name: String, detail: String? = nil, steps: [PlanStep]) {
        self.name = name
        self.detail = detail
        self.steps = steps
    }

    var totalRows: Int { steps.compactMap(\.rows).reduce(0, +) }
}

/// Everything the calculator works out for one project.
struct ProjectPlan: Identifiable, Equatable {
    var id: UUID = UUID()
    var title: String
    var gauge: Gauge
    var facts: [PlanFact]
    var sections: [PlanSection]
    /// Stitch blocks feeding the yarn estimate.
    var blocks: [StitchBlock]
    var notes: [String] = []
    /// Things that are off: gauge implausible, measurements inconsistent.
    var warnings: [String] = []
    /// Finished measurements the schematic is drawn from.
    var metrics: PlanMetrics = PlanMetrics()

    var totalStitches: Double { blocks.reduce(0) { $0 + $1.stitches } }
    var totalRows: Int { sections.reduce(0) { $0 + $1.totalRows } }

    func metres() -> Double { YarnEstimator.metres(for: blocks, gauge: gauge) }

    /// Flat plain-text rendering, used for sharing and export.
    func plainText(units: UnitSystem) -> String {
        var lines: [String] = [title, String(repeating: "=", count: title.count), ""]
        lines.append("Gauge: \(gauge.describe(in: units))")
        lines.append("")
        for fact in facts {
            if let detail = fact.detail {
                lines.append("\(fact.label): \(fact.value)  (\(detail))")
            } else {
                lines.append("\(fact.label): \(fact.value)")
            }
        }
        for section in sections {
            lines.append("")
            lines.append(section.name.uppercased())
            if let detail = section.detail { lines.append(detail) }
            for step in section.steps {
                var line = "  • \(step.text)"
                if let count = step.stitchCount { line += " (\(count) sts)" }
                lines.append(line)
            }
        }
        if !notes.isEmpty {
            lines.append("")
            lines.append("NOTES")
            notes.forEach { lines.append("  • \($0)") }
        }
        if !warnings.isEmpty {
            lines.append("")
            lines.append("CHECK THESE")
            warnings.forEach { lines.append("  ! \($0)") }
        }
        return lines.joined(separator: "\n")
    }
}

/// Common body measurements a project is calculated from. Stored in cm.
struct BodyMeasurements: Codable, Equatable {
    var chest: Double = 100
    var headCircumference: Double = 56
    var neckCircumference: Double = 38
    var upperArm: Double = 34
    var wrist: Double = 20
    var bodyLength: Double = 62
    var sleeveLength: Double = 45
    var footCircumference: Double = 23
    var footLength: Double = 24
    var handCircumference: Double = 20
    var handLength: Double = 19

    /// Standard adult sizes, so a knitter who has not measured can start.
    static let sizePresets: [(name: String, measurements: BodyMeasurements)] = [
        ("Child 6–8", BodyMeasurements(
            chest: 66, headCircumference: 51, neckCircumference: 30, upperArm: 22,
            wrist: 15, bodyLength: 42, sleeveLength: 33,
            footCircumference: 17, footLength: 18, handCircumference: 15, handLength: 14)),
        ("Adult XS", BodyMeasurements(
            chest: 82, headCircumference: 54, neckCircumference: 34, upperArm: 27,
            wrist: 16, bodyLength: 56, sleeveLength: 43,
            footCircumference: 20, footLength: 22, handCircumference: 17, handLength: 17)),
        ("Adult S", BodyMeasurements(
            chest: 90, headCircumference: 55, neckCircumference: 36, upperArm: 30,
            wrist: 17, bodyLength: 59, sleeveLength: 44,
            footCircumference: 21, footLength: 23, handCircumference: 18, handLength: 18)),
        ("Adult M", BodyMeasurements()),
        ("Adult L", BodyMeasurements(
            chest: 110, headCircumference: 57, neckCircumference: 40, upperArm: 37,
            wrist: 21, bodyLength: 65, sleeveLength: 46,
            footCircumference: 25, footLength: 26, handCircumference: 22, handLength: 20)),
        ("Adult XL", BodyMeasurements(
            chest: 122, headCircumference: 58, neckCircumference: 42, upperArm: 41,
            wrist: 22, bodyLength: 68, sleeveLength: 47,
            footCircumference: 27, footLength: 27, handCircumference: 24, handLength: 21)),
    ]
}
